### ISSUE TITLE
Unexpected error when using generic key record lookup and `Parameters`

### ISSUE THREAD

**Issue by OliverJAsh** [microsoft/TypeScript#48484]
# Bug Report

### 🔎 Search Terms

generics
mapped type
keyof
parameters
function
rest

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about generics

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/C4TwDgpgBAsghmAShAxgewE4BMA8AVKCAD2AgDssBnKZdbHS4DASzIHMAaKACgDp+wcDHAC2lAFxQ4ZEAG0AugEooAXgB8UmWo0qoAbwBQUY1AD0pqIFByKMAAWzagHdMAa0pGT5qLIDSUVlAuECBoAGZQePKSeL7yANwGHsZegDLkNvbUWGgQlGQA5MBJ3n4BQSHhkZJ8AkKiElAACrUiEKQYlPixasrqNK0ArhhkeOAQnT7yagkAvgkGoJBQAPowILSYWKr6RQCCVT0ajCzsCSZQAEL7qodMrGynJgDCV71HdzMJC9DwYJBY69gtj8AbgVmtUBspoksKgADZCaDoMiMGySH5-EFzJEo0JkLY4PzEUgUahlMKwBAYiHYNTcFySHwHbZnLw7DBsfotMjAKDkr5QPIKPL+ahkNA8uCUSjMNhkOAAI1h0GAaCggmELTavPC-LyTQ1rQg7RwYJBXTyRWAshc8m4ig+BiAA)

### 💻 Code

```ts
type MapRecord<T extends Record<string, (...params: any[]) => any>> = {
    // ✅ this works
    // [K in keyof T]: T[K];

    // ❌ this doesn't
    [K in keyof T]: (...params: Parameters<T[K]>) => ReturnType<T[K]>;
};

type _MyRecord = {
    A: () => string;
    B: () => string;
    C: () => string;
};
type MappedRecord = MapRecord<_MyRecord>;

declare const t: MappedRecord;

const fn = <K extends keyof MappedRecord>(k: K) => {
    // Argument of type '[]' is not assignable to parameter of type 'Parameters<_MyRecord[K]>'
    t[k]();
};
```

### 🙁 Actual behavior

No error.

### 🙂 Expected behavior

Error (see code comments above).

**Comment by encendre** [comment]
I'm not sure if it's related but it looks like but i have the same kind of problem with `ReturnType`

### Code: 

```ts
type MethodMap = Record<string, (...args: unknown[]) => unknown>

export class Router<
  M extends MethodMap
> {
  constructor(private methodMap: M) {
  }

  handle<K extends keyof M> (method: K, params: Parameters<M[K]>): ReturnType<M[K]> {
    const f = this.methodMap[method]
    const result = f(params)

    return result
  }
}
```

[Playground Link](https://www.typescriptlang.org/play?ts=4.6.2#code/C4TwDgpgBAshwAsD2ATGBDMUC8UBKEAxkgE4oA8AzsCQJYB2A5gDRQAUAdF+iY5QFxQArvQDW9JAHd6AbQC6AShwA+YWInTlAKC0QAHmFLAohADbpKlfEiHAIJclqiwo+u-RRW4iVBjBbVAG8nEyR6ahIhQmBSNjA6ADd0OygAW3hkNExBGCVg5wBfHWcEdA9TCHIAaVc9d08oUQgQJAAzWFU2dJ8UQSrWMB50VIEoAAUh7vtKchgZKrllBUECYCESegAVcEq5haCQ52Jw43bcRFpKDm7Mvxkb1DlD0JOoEghKIVNjXFa4ycoCmKzje8HW9FBn2+ISKRSAA)

note that with `any` instead of `unknown` there is no typescript error but the intermediate value `result` is of type `any` and is therefore not inferred correctly

**Comment by ahejlsberg** [comment]
I'm going to say this is working as intended. Following #47109 we do a better job of preserving the type of `t[k]` in a call such that non-uniform scenarios like this are supported:

```ts
type MapRecord<T extends Record<string, (...params: any[]) => any>> = {
    [K in keyof T]: (...params: Parameters<T[K]>) => ReturnType<T[K]>;
};

type _MyRecord = {
    A: () => string;
    B: (x: number) => string;
    C: (a: string, b: string) => string;
};

type MappedRecord = MapRecord<_MyRecord>;

declare const t: MappedRecord;

const fn = <K extends keyof MappedRecord>(k: K, params: Parameters<_MyRecord[K]>) => {
    t[k](...params);
};
```

Previously we'd only permit calls when the parameter list is valid for all methods in `_MyRecord`. However, now we fail to recognize the (likely less common) scenario where the methods _do_ have the same parameter list. Would be nice to do better there, but I think the new scenario we enable is more important.
